File: Runtime/TransitionTarget.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.StochasticTime

Friend Class TransitionTarget
    Inherits STSimDistributionBase

    Private m_TransitionGroupId As Integer
    Private m_ExpectedAmount As Double
    Private m_Multiplier As Double = 1.0

    Public Sub New(
        ByVal iteration As Nullable(Of Integer),
        ByVal timestep As Nullable(Of Integer),
        ByVal stratumId As Nullable(Of Integer),
        ByVal secondaryStratumId As Nullable(Of Integer),
        ByVal transitionGroupId As Integer,
        ByVal targetAmount As Nullable(Of Double),
        ByVal distributionTypeId As Nullable(Of Integer),
        ByVal distributionFrequency As Nullable(Of DistributionFrequency),
        ByVal distributionSD As Nullable(Of Double),
        ByVal distributionMin As Nullable(Of Double),
        ByVal distributionMax As Nullable(Of Double))

        MyBase.New(
            iteration,
            timestep,
            stratumId,
            secondaryStratumId,
            targetAmount,
            distributionTypeId,
            distributionFrequency,
            distributionSD,
            distributionMin,
            distributionMax)

        Me.m_TransitionGroupId = transitionGroupId

    End Sub

    Public ReadOnly Property TransitionGroupId As Integer
        Get
            Return Me.m_TransitionGroupId
        End Get
    End Property

    Public Property ExpectedAmount As Double
        Get
            Return Me.m_ExpectedAmount
        End Get
        Set(ByVal value As Double)
            Me.m_ExpectedAmount = value
        End Set
    End Property

    Public Property Multiplier As Double
        Get
            Return Me.m_Multiplier
        End Get
        Set(ByVal value As Double)
            Me.m_Multiplier = value
        End Set
    End Property

    Public Overrides Function Clone() As STSimDistributionBase

        Dim t As New TransitionTarget(
             Me.Iteration,
             Me.Timestep,
             Me.StratumId,
             Me.SecondaryStratumId,
             Me.TransitionGroupId,
             Me.DistributionValue,
             Me.DistributionTypeId,
             Me.DistributionFrequency,
             Me.DistributionSD,
             Me.DistributionMin,
             Me.DistributionMax)

        t.ExpectedAmount = Me.ExpectedAmount
        t.Multiplier = Me.Multiplier

        Return t

    End Function

End Class

File: Runtime/TransitionTarget.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System.Diagnostics;
using System.Collections.Generic;
using SyncroSim.Common;
using SyncroSim.StochasticTime;

namespace SyncroSim.STSim
{
    internal class TransitionTarget : STSimDistributionBase
    {
        private int m_TransitionGroupId;
        private double m_ExpectedAmount;
        private double m_Multiplier = 1.0;
        private List<TransitionTargetPrioritization> m_Prioritizations;
        private MultiLevelKeyMap4<TransitionTargetPrioritization> m_PrioritizationMap;

        public TransitionTarget(
            int? iteration, int? timestep, int? stratumId, int? secondaryStratumId, int? tertiaryStratumId, 
            int transitionGroupId, double? targetAmount, int? distributionTypeId, DistributionFrequency? distributionFrequency,
            double? distributionSD, double? distributionMin, double? distributionMax) : base(iteration, timestep, stratumId, secondaryStratumId, 
                tertiaryStratumId, targetAmount, distributionTypeId, distributionFrequency, distributionSD, distributionMin, distributionMax)
        {

            this.m_TransitionGroupId = transitionGroupId;
        }

        public int TransitionGroupId
        {
            get
            {
                return this.m_TransitionGroupId;
            }
        }

        public double ExpectedAmount
        {
            get
            {
                return this.m_ExpectedAmount;
            }
            set
            {
                this.m_ExpectedAmount = value;
            }
        }

        public double Multiplier
        {
            get
            {
                return this.m_Multiplier;
            }
            set
            {
                this.m_Multiplier = value;
            }
        }

        public List<TransitionTargetPrioritization> Prioritizations
        {
            get
            {
                return m_Prioritizations;
            }
        }

        public override STSimDistributionBase Clone()
        {
            TransitionTarget t = new TransitionTarget(
                this.Iteration, this.Timestep, this.StratumId, 
                this.SecondaryStratumId, this.TertiaryStratumId, this.TransitionGroupId, 
                this.DistributionValue, this.DistributionTypeId, this.DistributionFrequency, 
                this.DistributionSD, this.DistributionMin, this.DistributionMax);

            t.ExpectedAmount = this.ExpectedAmount;
            t.Multiplier = this.Multiplier;

            return t;
        }

        public void SetPrioritizations(List<TransitionTargetPrioritization> prioritizations)
        {
            this.ClonePrioritizationList(prioritizations);
            this.CreatePrioritizationMap();
        }

        public TransitionTargetPrioritization GetPrioritization(
            int stratumId,
            int? secondaryStratumId,
            int? tertiaryStratumId,
            int stateClassId)
        {
            return this.m_PrioritizationMap.GetItem(
                stratumId, secondaryStratumId, tertiaryStratumId, stateClassId);
        }

        private void ClonePrioritizationList(List<TransitionTargetPrioritization> prioritizations)
        {
            Debug.Assert(this.m_Prioritizations == null);
            this.m_Prioritizations = new List<TransitionTargetPrioritization>();

            foreach (TransitionTargetPrioritization t in prioritizations)
            {
                this.m_Prioritizations.Add(new TransitionTargetPrioritization(
                    t.Iteration, 
                    t.Timestep, 
                    t.TransitionGroupId,
                    t.StratumId, 
                    t.SecondaryStratumId, 
                    t.TertiaryStratumId, 
                    t.StateClassId, 
                    t.Priority));
            }

            Debug.Assert(this.m_Prioritizations.Count == prioritizations.Count);
        }

        private void CreatePrioritizationMap()
        {
            Debug.Assert(this.m_PrioritizationMap == null);
            this.m_PrioritizationMap = new MultiLevelKeyMap4<TransitionTargetPrioritization>();

            foreach (TransitionTargetPrioritization pri in this.m_Prioritizations)
            {
                TransitionTargetPrioritization p = this.m_PrioritizationMap.GetItemExact(
                    pri.StratumId, pri.SecondaryStratumId, pri.TertiaryStratumId, pri.StateClassId);

                if (p == null)
                {
                    this.m_PrioritizationMap.AddItem(
                        pri.StratumId,
                        pri.SecondaryStratumId,
                        pri.TertiaryStratumId,
                        pri.StateClassId,
                        pri);
                }
            }
        }
    }
}
